pint_server/data_update.py

```diff
@@ -207,20 +214,39 @@ def orm_update_table(db, provider, table_name, table_rows, version):
                  repr(model.__name__), repr(provider),
                  repr(table_name))
 
+    # fields used to identify a matching existing entry
+    identity_fields = IDENTITY_OVERRIDES.get(model.__name__, [])
+    if identity_fields:
+        LOG.debug("For %s using %s instead of primary key fields for "
+                  "existing entry checking", repr(model.__name__),
+                  repr(identity_fields))
+
     LOG.debug("Attempting to add %d new entries for model %s",
                  len(table_rows), repr(model.__name__))
     rows_added = 0
     rows_updated = 0
     for row_data in table_rows:
-        search_data = {k:v for k, v in row_data.items()
-                            if getattr(model.__table__.columns,
-                                       k).primary_key}
-
-        # If table is using a surrogate integer index as the primary key
-        # then just search for an exact match
+        # if identity_fields overrides have been specified use those
+        # fields to construct the search_data that will be used to
+        # check for existing matching entries, otherwise use the
+        # fields that are marked as primary keys.
+        if identity_fields:
+            search_data = {k:row_data[k] for k in identity_fields}
+        else:
+            search_data = {k:v for k, v in row_data.items()
+                               if getattr(model.__table__.columns,
+                                          k).primary_key}
+
+        # if no search_data was identified, fall back on doing a whole
+        # row match search, to see if there is already an exact match
+        # present.
         if not search_data:
```

Comment:
search_data should always have something right?

Reply:
Nope.

If no identity overrides were specified, and no columns are marked as primary keys in the table definition, then search_data can be an empty list at this point, hence assigning row_data to it so that we can do a search for an exact match, which works fine for identifying existing servers tables entries for now, but could be an issue if we need to update any servers table entries with new data before we stop using this script...

This was basically the issue we had to deal with as the `microsoftimages` table has no columns marked as primary keys under the new schema; instead it has a unique id sequence column.

As a result search_data was empty, so we were falling back on the exact match search, finding no matches, and thus adding new entries, instead of updating the existing entry as should have been the case.